pkg/ingester/ingester.go

```diff
@@ -0,0 +1,126 @@
+package ingester
+
+import (
+	"errors"
+	"fmt"
+	"github.com/mitlibraries/mario/pkg/client"
+	"github.com/mitlibraries/mario/pkg/consumer"
+	"github.com/mitlibraries/mario/pkg/generator"
+	"github.com/mitlibraries/mario/pkg/pipeline"
+	"github.com/mitlibraries/mario/pkg/transformer"
+	"io"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+)
+
+type Config struct {
+	Filename  string
+	Source    string
+	Consumer  string
+	Url       string
+	Index     string
+	V4        bool
+	Prefix    string
+	Promote   bool
+	Rulesfile string
+}
+
+func NewStream(filename string) (io.ReadCloser, error) {
+	parts, err := url.Parse(filename)
+	if err != nil {
+		return nil, err
+	}
+	if parts.Scheme == "s3" {
+		return client.GetS3Obj(parts.Host, parts.Path)
+	} else {
+		return os.Open(filename)
+	}
+}
+
+type Ingester struct {
+	Stream    io.ReadCloser
+	config    Config
+	generator pipeline.Generator
+	consumer  pipeline.Consumer
+	Client    client.Indexer
+}
+
+func (i *Ingester) Configure(config Config) error {
+	var err error
+	// Configure generator
+	if config.Source == "json" {
+		i.generator = &generator.JSONGenerator{i.Stream}
+	} else if config.Source == "marc" {
+		i.generator = &generator.MarcGenerator{
+			Marcfile:  i.Stream,
+			Rulesfile: config.Rulesfile,
+		}
+	} else if config.Source == "archives" {
+		i.generator = &generator.ArchivesGenerator{Archivefile: i.Stream}
+	} else {
+		return errors.New("Unknown source data")
+	}
+
+	// Configure consumer
+	if config.Consumer == "es" {
+		if config.Index == "" {
+			if strings.Contains(config.Filename, "mit01_edsu1") {
```

Comment:
Not blocking comment: This business logic could likely use a comment explaining it (likely not a new concern with this PR). Explaining we are detecting and update file and thus it requires an existing index to update into will likely save one of us a lot of confusion in the future.